src/preact/component/intersection-observer-resize.js

```diff
@@ -0,0 +1,50 @@
+import {toWin} from '#core/window';
+
+import {useEffect} from '#preact';
+
+const ioForWindow = new WeakMap();
+const callbackMap = new WeakMap();
+
+/**
+ * Uses a shared IntersectionObserver per window instance to observe the given `ref`.
+ *
+ * @param {{current: Element}} ref
+ * @param {function(IntersectionObserverEntry)} callback
+ * @param {?Window} targetWin
+ */
+export function useIntersectionObserver(ref, callback, targetWin = null) {
+  useEffect(() => {
+    const node = ref.current;
+    if (!node) {
+      return;
+    }
+    // If a window instance is not provided, uses the one local to the given `ref`.
+    const win = targetWin ?? toWin(node.ownerDocument?.defaultView);
+    if (!win) {
+      return;
+    }
+    callbackMap.set(node, callback);
+
+    if (ioForWindow.has(win)) {
+      ioForWindow.get(win).observe(node);
+    } else {
+      const io = new win.IntersectionObserver((entries) => {
+        entries.reduceRight((accumulator, entry) => {
+          const {target} = entry;
+          if (!accumulator.has(target)) {
+            accumulator.add(target);
+            callbackMap.get(target)(entry);
+          }
+          return accumulator;
+        }, new Set());
+      });
+      ioForWindow.set(win, io);
+      io.observe(node);
+    }
+
+    return () => {
+      ioForWindow.get(win).unobserve(node);
+      callbackMap.delete(node);
+    };
+  }, [callback, ref, targetWin]);
```

Comment:
Unsure if bug or not: does using a `ref` work in the dep array vs. using `ref.current`?

Reply:
Another good point. Based on this [blog post](https://medium.com/@teh_builder/ref-objects-inside-useeffect-hooks-eb7c15198780) we need to change this to a callback ref. WIP.